rotkehlchen/db/drivers/gevent.py

```diff
@@ -295,6 +313,87 @@ def write_ctx(self) -> Generator['DBCursor', None, None]:
 
+    @contextmanager
+    def savepoint_ctx(
+            self,
+            savepoint_name: Optional[str] = None,
+    ) -> Generator['DBCursor', None, None]:
+        """
+        Creates a savepoint context with the provided name. If the code inside the savepoint fails,
+        rolls back this savepoint, otherwise releases it (aka forgets it -- this is not commited to the DB).  # noqa: E501
+        Savepoints work like nested transactions, more information here: https://www.sqlite.org/lang_savepoint.html  # noqa: E501
+        """
+        cursor, savepoint_name = self.enter_savepoint(savepoint_name)
+        try:
+            yield cursor
+        except Exception:
+            self.rollback_savepoint(savepoint_name)
+            raise
+        else:
+            self.release_savepoint(savepoint_name)
+        finally:
+            cursor.close()  # lgtm [py/should-use-with]
+
+    def enter_savepoint(self, savepoint_name: Optional[str] = None) -> Tuple['DBCursor', str]:
+        """
+        Creates an sqlite savepoint with the given name. If None is given, a uuid is created.
+        Returns cursor and savepoint's name.
+        May raise:
+        - ContextError if a savepoint with the same name already exists. Can only happen in case of
+        manually specified name.
+        """
+        if savepoint_name is None:
+            savepoint_name = str(uuid4())
+        if savepoint_name in self.savepoints:
+            raise ContextError(
+                f'Wanted to enter savapoint {savepoint_name} but a savepoint with the same name '
+                f'already exists. Current savepoints: {list(self.savepoints)}',
+            )
+        cursor = self.cursor()
+        cursor.execute(f'SAVEPOINT "{savepoint_name}"')
+        self.savepoints[savepoint_name] = None
+        return cursor, savepoint_name
+
+    def _modify_savepoint(
+            self,
+            rollback_or_release: Literal['rollback', 'release'],
+            savepoint_name: Optional[str],
+    ) -> None:
+        if len(self.savepoints) == 0:
+            raise ContextError(
+                f'Incorrect use of savepoints! Wanted to rollback to savepoint {savepoint_name}, '
+                f'but the stack is empty.',
+            )
+        list_savepoints = list(self.savepoints)
+        if savepoint_name is None:
+            savepoint_name = list_savepoints[-1]
+        elif savepoint_name not in self.savepoints:
+            raise ContextError(
+                f'Incorrect use of savepoints! Wanted to revert savepoint {savepoint_name}, '
+                f'but it is not present in the stack: {list_savepoints}',
+            )
+        if rollback_or_release == 'rollback':
+            self.execute(f'ROLLBACK TO SAVEPOINT "{savepoint_name}"')
+        else:  # release
+            self.execute(f'RELEASE SAVEPOINT "{savepoint_name}"')
+        self.savepoints = dict.fromkeys(list_savepoints[:list_savepoints.index(savepoint_name)])
```

Comment:
Uhmmm, I think this will not work. Why not just `del self.savepoints[savepoint_name]`?

Reply:
Oh nevermind, I think I understand the way it works now.

Reply:
Yeah, will add a comment in the code also